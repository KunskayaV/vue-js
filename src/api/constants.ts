import {
  DATE_PROPERTY,
  ESearchByFilter,
  ESortByValues,
  GENRE_PROPERTY,
  RATING_PROPERTY,
  TITLE_PROPERTY
} from '@/constants'

export const BASE_URL = 'https://tame-erin-pike-toga.cyclic.app'

export const searchParamsMap = {
  [ESearchByFilter.Genre]: GENRE_PROPERTY,
  [ESearchByFilter.Title]: TITLE_PROPERTY
}

export const sortParamsMap = {
  [ESortByValues.Date]: DATE_PROPERTY,
  [ESortByValues.Rating]: RATING_PROPERTY
}

export const PARTIAL_MATCH_TAIL = '_like'
export const SORT_KEY = '_sort'
export const ORDER_KEY = '_order'
