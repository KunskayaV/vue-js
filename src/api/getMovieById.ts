import type { TMovie } from '@/types'
import apiClient from './apiClient'

export async function getMovieById(id: number): Promise<TMovie> {
  const response = await apiClient.get(`/movies/${id}`)

  return response.data
}
