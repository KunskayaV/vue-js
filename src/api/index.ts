export * from './requests/getMovies'
export * from './requests/getMovieById'
