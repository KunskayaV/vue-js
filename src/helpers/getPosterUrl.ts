export function getPosterUrl(): string {
  const random = Math.floor(Math.random() * 567)
  const lock = Math.floor(Math.random() * 132)

  return `https://loremflickr.com/g/320/240/poster,movie,film?random=${random}&lock=${lock}`
}
