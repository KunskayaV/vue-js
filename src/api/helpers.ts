import { PARTIAL_MATCH_TAIL } from './constants'

export function getPartialMatchingParam(param: string): string {
  return `${param}${PARTIAL_MATCH_TAIL}`
}

export function removePartialMatchingTail(param: string): string {
  return param.replace(PARTIAL_MATCH_TAIL, '')
}
