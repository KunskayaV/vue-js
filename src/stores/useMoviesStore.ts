import {
  DATE_PROPERTY,
  ESearchByFilter,
  ESortByValues,
  GENRE_PROPERTY,
  TITLE_PROPERTY,
  RATING_PROPERTY,
  MATCH_MOVIES_LENGTH
} from '@/constants'
import { defineStore } from 'pinia'
import { computed, ref } from 'vue'
import { getMovies } from '@/api'
import type { TMovie } from '@/types'

export const useMoviesStore = defineStore('moviesStore', () => {
  const searchValue = ref('')
  const searchByValue = ref(ESearchByFilter.Title)
  const sortByValue = ref(ESortByValues.Date)
  const moviesData = ref<TMovie[]>([])
  const isMoviesDataLoaded = ref(false)

  function setSearchValue(value: string) {
    searchValue.value = value
  }
  function setSearchByValue(value: ESearchByFilter) {
    searchByValue.value = value
  }
  function setSortByValue(value: ESortByValues) {
    sortByValue.value = value
  }

  async function getMoviesData() {
    const data = await getMovies()

    moviesData.value = data
    isMoviesDataLoaded.value = true
  }

  function getMovieDetailsById(id: number): TMovie | undefined {
    return moviesData.value.find((movie) => movie.id === id)
  }

  const filteredMovies = computed(() => {
    return moviesData.value.filter((movie) => {
      if (searchByValue.value == ESearchByFilter.Title) {
        return movie[TITLE_PROPERTY].toLowerCase().includes(searchValue.value.toLowerCase())
      } else {
        return movie[GENRE_PROPERTY].some((genre) =>
          genre.toLowerCase().includes(searchValue.value.toLowerCase())
        )
      }
    })
  })

  const movies = computed(() => {
    return [...filteredMovies.value].sort((a, b) => {
      const property = sortByValue.value == ESortByValues.Rating ? RATING_PROPERTY : DATE_PROPERTY

      if (a[property] < b[property]) {
        return 1
      } else {
        return -1
      }
    })
  })

  function getMatchMovies(id: number): TMovie[] {
    let index = -1
    const matchMovies = []
    const baseMovie = movies.value.find((movie) => movie.id === id)

    while (matchMovies.length < MATCH_MOVIES_LENGTH && ++index < movies.value.length) {
      const movieToCheck = movies.value[index]

      if (baseMovie && movieToCheck.genres.includes(baseMovie?.genres[0])) {
        matchMovies.push(movieToCheck)
      }
    }

    return matchMovies
  }

  return {
    searchValue,
    searchByValue,
    sortByValue,
    movies,
    setSearchValue,
    setSearchByValue,
    setSortByValue,
    getMoviesData,
    getMovieDetailsById,
    isMoviesDataLoaded,
    getMatchMovies
  }
})
