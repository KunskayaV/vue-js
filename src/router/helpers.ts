import { getPartialMatchingParam } from '@/api/helpers'
import { GENRE_PROPERTY, TITLE_PROPERTY } from '@/constants'
import type { LocationQuery } from 'vue-router'

export function removeSearchByFromQuery(query: LocationQuery): LocationQuery {
  const newQuery = { ...query }
  delete newQuery[getPartialMatchingParam(GENRE_PROPERTY)]
  delete newQuery[getPartialMatchingParam(TITLE_PROPERTY)]

  return newQuery
}

export function getSearchByKeyFromQuery(query: LocationQuery): string {
  if (getPartialMatchingParam(GENRE_PROPERTY) in query) {
    return GENRE_PROPERTY
  } else {
    return TITLE_PROPERTY
  }
}
