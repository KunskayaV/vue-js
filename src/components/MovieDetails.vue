<template>
  <div>
    <div v-if="!movieDetails" class="loading-container">
      <span>LOADING...</span>
    </div>
    <div v-else class="movie-details">
      <img
        class="poster-image"
        :src="posterUrl"
        :alt="movieDetails.storyline"
        width="300"
        height="400"
      />
      <div class="movie-description">
        <MovieDetailsTitle
          :title="movieDetails.title"
          :rating="movieDetails.averageRating"
          :genre="getItemGenreLine(movieDetails?.genres)"
        />
        <MovieDetailsDuration :year="movieDetails.year" :duration="movieDetails.duration" />
        <p v-if="!!movieDetails.storyline">{{ movieDetails.storyline }}</p>
      </div>
    </div>
  </div>
</template>

<script setup lang="ts">
import MovieDetailsTitle from './MovieDetailsTitle.vue'
import MovieDetailsDuration from './MovieDetailsDuration.vue'
import { useMoviesStore } from '@/stores/useMoviesStore'
import { getPosterUrl } from '@/helpers/getPosterUrl'
import { getItemGenreLine } from '@/helpers/getItemGenreLine'
import { computed } from 'vue'

// TODO: temp solution for id
const props = defineProps<{ id: number | undefined }>()

const appState = useMoviesStore()
const movieDetails = computed(() => props.id && appState.getMovieDetailsById(props.id))

const posterUrl = getPosterUrl()
</script>

<style scoped>
.movie-details {
  display: flex;
  flex-direction: row;
  align-items: center;
  padding-top: 16px;
  padding-bottom: 48px;
}

.movie-description {
  margin-left: 24px;
  flex: 1;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  padding-inline: 24px;
  color: #ffffff;
}

.loading-container {
  display: flex;
  justify-content: center;
  padding-top: 5%;
  height: 20%;
  font-size: 1.5rem;
  line-height: 1.2;
  letter-spacing: 1.5;
  color: white;
}

.poster-image {
  background-color: #bdbdbd;
}
</style>
