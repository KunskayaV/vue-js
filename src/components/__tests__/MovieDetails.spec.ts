import { describe, it, expect, afterAll, beforeEach, vi } from 'vitest'
import { setActivePinia, createPinia } from 'pinia'

import { mount } from '@vue/test-utils'
import MovieDetails from '@/components/MovieDetails.vue'
import { mockMovie } from '@/__mocks__'
import { useMoviesStore } from '@/stores/useMoviesStore'
import * as api from '@/api'
import * as urlHelpers from '@/helpers/getPosterUrl'
import { nextTick } from 'vue'

describe('MovieDetails', () => {
  beforeEach(() => {
    vi.spyOn(api, 'getMovieById').mockResolvedValue(mockMovie)
    vi.spyOn(urlHelpers, 'getPosterUrl').mockReturnValue('url')
    setActivePinia(createPinia())
  })

  afterAll(() => {
    vi.restoreAllMocks()
  })

  it('should match initial empty snapshot', () => {
    const wrapper = mount(MovieDetails, {
      props: { id: mockMovie.id }
    })

    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.text()).toContain('LOADING...')
  })

  it.only('should match snapshot with data', async () => {
    const state = useMoviesStore()
    await state.getMoviesData()

    const wrapper = mount(MovieDetails, { props: { id: mockMovie.id } })

    await nextTick()

    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.text()).not.toContain('LOADING...')
  })
})
