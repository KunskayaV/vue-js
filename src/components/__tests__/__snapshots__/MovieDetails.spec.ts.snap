// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MovieDetails > should match snapshot with data 1`] = `
"<div data-v-b1321ac7=\\"\\">
  <div data-v-b1321ac7=\\"\\" class=\\"movie-details\\"><img data-v-b1321ac7=\\"\\" class=\\"poster-image\\" src=\\"url\\" alt=\\"Chronicles the experiences of a formerly successful banker as a prisoner in the gloomy jailhouse of Shawshank after being found guilty of a crime he did not commit. The film portrays the man's unique way of dealing with his new, torturous life; along the way he befriends a number of fellow prisoners, most notably a wise long-term inmate named Red.                Written by
J-S-Golden\\" width=\\"300\\" height=\\"400\\">
    <div data-v-b1321ac7=\\"\\" class=\\"movie-description\\">
      <div data-v-8cdb4aee=\\"\\" data-v-b1321ac7=\\"\\" class=\\"title-container\\">
        <div data-v-8cdb4aee=\\"\\" class=\\"title-wrapper\\"><span data-v-8cdb4aee=\\"\\" class=\\"movie-title\\" date-testid=\\"details-title\\">The Shawshank Redemption</span><span data-v-8cdb4aee=\\"\\" class=\\"movie-rating\\" date-testid=\\"details-rating\\">0</span></div><span data-v-8cdb4aee=\\"\\" class=\\"movie-genre\\" date-testid=\\"details-genre\\">Crime &amp; Drama</span>
      </div>
      <div data-v-b9b942cb=\\"\\" data-v-b1321ac7=\\"\\" class=\\"duration-container\\"><span data-v-b9b942cb=\\"\\" date-testid=\\"details-year\\">1995</span><span data-v-b9b942cb=\\"\\" date-testid=\\"details-duration\\">142 min</span></div>
      <p data-v-b1321ac7=\\"\\">Chronicles the experiences of a formerly successful banker as a prisoner in the gloomy jailhouse of Shawshank after being found guilty of a crime he did not commit. The film portrays the man's unique way of dealing with his new, torturous life; along the way he befriends a number of fellow prisoners, most notably a wise long-term inmate named Red. Written by
        J-S-Golden</p>
    </div>
  </div>
</div>"
`;
